abci/README.md

```diff
@@ -0,0 +1,220 @@
+3. Each lane will select transactions from its mempool, verify them according to its own rules, and return a list of valid transactions to add and a list of transactions to remove i.e. see [`PrepareLane`](../block/base/abci.go).
+4. The transactions will only be added to the current proposal being built iff the transactions are under the block gas and size limit for that lane. If the transactions are over the limit, they will NOT be added to the proposal and the next lane will be called.
+5. Once all lanes have been called, a final proposal is returned with all of the valid transactions from each lane.
+
+If any of the lanes fail to `PrepareLane`, the next lane will be called and the proposal will be built from the remaining lanes. This is a fail-safe mechanism to ensure that the proposal is always built, even if one of the lanes fails to prepare. Additionally, state is mutated iff the lane is successful in preparing its portion of the proposal.
+
+To customize how much block-space a given lane consumes, you have to configure the `MaxBlockSpace` variable in your lane configuration object (`LaneConfig`). Please visit [`lanes.go`](../tests/app/lanes.go) for an example. This variable is a map of lane name to the maximum block space that lane can consume. Note that if the Block SDK module is utilized, the `MaxBlockSpace` variable will be overwritten by the governance configured value.
+
+### Proposal Construction Example
+
+> Let's say your application has 3 lanes, `A`, `B`, and `C`, and the order of lanes is `A -> B -> C`. 
+> 
+> * Lane `A` has a `MaxBlockSpace` of 1000 bytes and 500 gas limit.
+> * Lane `B` has a `MaxBlockSpace` of 2000 bytes and 1000 gas limit.
+> * Lane `C` has a `MaxBlockSpace` of 3000 bytes and 1500 gas limit.
+
+Lane `A` currently contains 4 transactions:
+
+* Tx1: 100 bytes, 100 gas
+* Tx2: 800 bytes, 300 gas
+* Tx3: 200 bytes, 100 gas
+* Tx4: 100 bytes, 100 gas
+
+Lane `B` currently contains 4 transactions:
+
+* Tx5: 1000 bytes, 500 gas
+* Tx6: 1200 bytes, 600 gas
+* Tx7: 1500 bytes, 300 gas
+* Tx8: 1000 bytes, 400 gas
+
+Lane `C` currently contains 4 transactions:
+
+* Tx9: 1000 bytes, 500 gas
+* Tx10: 1200 bytes, 600 gas
+* Tx11: 1500 bytes, 300 gas
+* Tx12: 100 bytes, 400 gas
+
+Assuming all transactions are valid according to their respective lanes, the proposal will be built as follows:
+
+```golang
+// Somewhere in abci.go a new proposal is created:
+blockProposal := proposals.NewProposal(...)
+...
+// First lane to be called is lane A, this will return the following transactions to add after PrepareLane is called:
+partialProposalFromLaneA := {
+    Tx1, // 100 bytes, 100 gas
+    Tx2, // 800 bytes, 300 gas
+    Tx3, // 200 bytes, 100 gas
+}
+...
+// First Lane A will update the proposal.
+if err := blockProposal.Update(partialProposalFromLaneA); err != nil {
+    return err
+}
+...
+// Next, lane B will be called with the following transactions to add after PrepareLane is called:
```

Comment:
We should note some thing like:

"Here, Tx6 is excluded because it and Tx5's gas would exceed the lane gas limit.  Tx7 is similarly excluded because it and Tx5's  size would exceed the lane block size limit.  Note that Tx5 will always be included first because it is ordered first and it is valid given the lane's constraints"